### ISSUE TITLE
ntopng --shutdown-when-done doesn't stop the process

### ISSUE THREAD

**Issue by gianlucapiccolo** [ntop/ntopng#4729]
### Use case
There is a local startup.lua that must be run at the system installation/upgrade to create an API user in ntopng. **Only the root user can execute that script** so at the start of that we check if it is run as root and if not nothing is done (the system ntopng service doesn't run as root).

### Solution
ntopng -v 2 **-U root --dont-change-user --shutdown-when-done** -i <(echo) -r "127.0.0.1" --http-port 31826

### Contest
On ntopng 4.0 the command was working fine. It exits successfully after displaying the following output:
```
16/Nov/2020 11:40:48 [Prefs.cpp:1106] ntopng will use redis 127.0.0.1@0
16/Nov/2020 11:40:48 [Ntop.cpp:2291] Setting local networks to 127.0.0.0/8
16/Nov/2020 11:40:48 [Redis.cpp:157] Successfully connected to redis 127.0.0.1:6379@0
16/Nov/2020 11:40:48 [Redis.cpp:157] Successfully connected to redis 127.0.0.1:6379@0
16/Nov/2020 11:40:48 [NtopPro.cpp:293] [LICENSE] Reading license from Redis
16/Nov/2020 11:40:48 [NtopPro.cpp:421] [LICENSE] Unable to validate license [Empty license file]
16/Nov/2020 11:40:48 [NtopPro.cpp:488] WARNING: [LICENSE] Invalid license [Empty license file]
16/Nov/2020 11:40:48 [NtopPro.cpp:505] WARNING: [LICENSE] ntopng will now run in Enterprise L edition for 10 minutes
16/Nov/2020 11:40:48 [NtopPro.cpp:507] WARNING: [LICENSE] before returning to community mode
16/Nov/2020 11:40:48 [NtopPro.cpp:509] WARNING: [LICENSE] You can buy a permanent license at http://shop.ntop.org
16/Nov/2020 11:40:48 [NtopPro.cpp:511] WARNING: [LICENSE] or run ntopng in community mode starting
16/Nov/2020 11:40:48 [NtopPro.cpp:512] WARNING: [LICENSE] ntopng --community
16/Nov/2020 11:40:48 [boot.lua:27] [delete_data_utils.lua:420] Deleting data [delete_if_redis_keys][OK]
16/Nov/2020 11:40:48 [boot.lua:27] [delete_data_utils.lua:420] Deleting data [delete_if_db][OK]
16/Nov/2020 11:40:48 [boot.lua:27] [delete_data_utils.lua:420] Deleting data [delete_if_ids][OK]
16/Nov/2020 11:40:48 [boot.lua:27] [delete_data_utils.lua:420] Deleting data [delete_if_data][OK]
16/Nov/2020 11:40:49 [Ntop.cpp:2396] Registered interface /dev/fd/63 [id: 0]
16/Nov/2020 11:40:49 [main.cpp:316] PID stored in file /var/run/ntopng.pid
16/Nov/2020 11:40:49 [Geolocation.cpp:150] Running without geolocation support.
16/Nov/2020 11:40:49 [Geolocation.cpp:151] To enable geolocation follow the instructions at
16/Nov/2020 11:40:49 [Geolocation.cpp:152] https://github.com/ntop/ntopng/blob/dev/doc/README.geolocation.md
16/Nov/2020 11:40:49 [HTTPserver.cpp:1371] Web server dirs [/usr/share/ntopng/httpdocs][/usr/share/ntopng/scripts]
16/Nov/2020 11:40:49 [HTTPserver.cpp:1374] HTTP server listening on 31826
16/Nov/2020 11:40:49 [main.cpp:386] Working directory: /var/lib/ntopng
16/Nov/2020 11:40:49 [main.cpp:388] Scripts/HTML pages directory: /usr/share/ntopng
16/Nov/2020 11:40:49 [Ntop.cpp:443] Welcome to ntopng x86_64 v.4.0.200827 - (C) 1998-20 ntop.org
16/Nov/2020 11:40:49 [NtopPro.cpp:692] [LICENSE] System Id:	BE150604920BAB21
16/Nov/2020 11:40:49 [NtopPro.cpp:693] [LICENSE] Edition:	Enterprise L
16/Nov/2020 11:40:49 [NtopPro.cpp:694] [LICENSE] License Type:	Time-Limited [Empty license file] License
16/Nov/2020 11:40:49 [NtopPro.cpp:714] [LICENSE] Validity:	Until Mon Nov 16 11:50:48 2020
16/Nov/2020 11:40:49 [PeriodicActivities.cpp:105] Started periodic activities loop...
16/Nov/2020 11:40:51 [startup.lua:197] Loading category lists...
16/Nov/2020 11:40:51 [startup.lua:199] Loading category lists done
16/Nov/2020 11:40:52 [startup.lua:210] Running /usr/share/ntopng/local/scripts/callbacks/system/startup.lua
16/Nov/2020 11:40:52 [startup.lua:211] [startup.lua:40] API user already existing...
16/Nov/2020 11:40:52 [startup.lua:214] Startup completed
16/Nov/2020 11:40:52 [PeriodicActivities.cpp:165] Each periodic activity script will use 2 threads
16/Nov/2020 11:40:52 [NetworkInterface.cpp:2352] Started packet polling on interface /dev/fd/63 [id: 0]...
16/Nov/2020 11:40:53 [PcapInterface.cpp:227] No more pcap files to read
16/Nov/2020 11:40:53 [PcapInterface.cpp:341] Terminated packet polling for /dev/fd/63
16/Nov/2020 11:40:55 [Ntop.cpp:2489] Terminating periodic activities
16/Nov/2020 11:40:55 [Ntop.cpp:2495] Executing shutdown script
16/Nov/2020 11:40:56 [ProtoStats.cpp:35] [IPv4]  0 B/0.00 Packets
16/Nov/2020 11:40:56 [ProtoStats.cpp:35] [IPv6]  0 B/0.00 Packets
16/Nov/2020 11:40:56 [ProtoStats.cpp:35] [ARP]   0 B/0.00 Packets
16/Nov/2020 11:40:56 [ProtoStats.cpp:35] [MPLS]  0 B/0.00 Packets
16/Nov/2020 11:40:56 [ProtoStats.cpp:35] [Other] 0 B/0.00 Packets
16/Nov/2020 11:40:56 [Ntop.cpp:2477] Polling shut down [interface: /dev/fd/63]
16/Nov/2020 11:40:56 [Ntop.cpp:2512] Deleted PID /var/run/ntopng.pid: [rc: 0][Success]
16/Nov/2020 11:40:56 [HTTPserver.cpp:1397] HTTP server terminated
16/Nov/2020 11:40:56 [AddressResolution.cpp:63] Address resolution stats [0 resolved][0 failures]
```

### The issue on 4.2
On ntopng 4.2 **the command remains appended forever** to the terminal and **it doesn't stop** anymore.
The following output is displayed:
```
16/Nov/2020 11:45:32 [Prefs.cpp:1126] ntopng will use redis 127.0.0.1@0
16/Nov/2020 11:45:32 [Ntop.cpp:2336] Setting local networks to 127.0.0.0/8,fe80::/10
16/Nov/2020 11:45:32 [Redis.cpp:157] Successfully connected to redis 127.0.0.1:6379@0
16/Nov/2020 11:45:32 [Redis.cpp:157] Successfully connected to redis 127.0.0.1:6379@0
16/Nov/2020 11:45:32 [NtopPro.cpp:297] [LICENSE] Reading license from Redis
16/Nov/2020 11:45:32 [NtopPro.cpp:425] [LICENSE] Unable to validate license [Empty license file]
16/Nov/2020 11:45:32 [NtopPro.cpp:493] WARNING: [LICENSE] Invalid license [Empty license file]
16/Nov/2020 11:45:32 [NtopPro.cpp:510] WARNING: [LICENSE] ntopng will now run in Enterprise L edition for 10 minutes
16/Nov/2020 11:45:32 [NtopPro.cpp:512] WARNING: [LICENSE] before returning to community mode
16/Nov/2020 11:45:32 [NtopPro.cpp:514] WARNING: [LICENSE] You can buy a permanent license at http://shop.ntop.org
16/Nov/2020 11:45:32 [NtopPro.cpp:516] WARNING: [LICENSE] or run ntopng in community mode starting
16/Nov/2020 11:45:32 [NtopPro.cpp:517] WARNING: [LICENSE] ntopng --community
16/Nov/2020 11:45:32 [boot.lua:31] [delete_data_utils.lua:425] Deleting data [delete_if_ids][OK]
16/Nov/2020 11:45:32 [boot.lua:31] [delete_data_utils.lua:425] Deleting data [delete_if_data][OK]
16/Nov/2020 11:45:32 [boot.lua:31] [delete_data_utils.lua:425] Deleting data [delete_if_redis_keys][OK]
16/Nov/2020 11:45:32 [boot.lua:31] [delete_data_utils.lua:425] Deleting data [delete_if_db][OK]
16/Nov/2020 11:45:32 [Ntop.cpp:2441] Registered interface /dev/fd/63 [id: 0]
16/Nov/2020 11:45:32 [main.cpp:312] PID stored in file /var/run/ntopng.pid
16/Nov/2020 11:45:32 [Geolocation.cpp:140] Loaded database dbip-asn-lite.mmdb [/usr/share/ntopng/httpdocs/geoip//dbip-asn-lite.mmdb][ip_version: 6]
16/Nov/2020 11:45:32 [Geolocation.cpp:140] Loaded database dbip-country-lite.mmdb [/usr/share/ntopng/httpdocs/geoip//dbip-country-lite.mmdb][ip_version: 6]
16/Nov/2020 11:45:32 [Geolocation.cpp:91] Using geolocation provided by DB-IP (https://db-ip.com)
16/Nov/2020 11:45:32 [HTTPserver.cpp:1529] Web server dirs [/usr/share/ntopng/httpdocs][/usr/share/ntopng/scripts]
16/Nov/2020 11:45:32 [HTTPserver.cpp:1532] HTTP server listening on 31826
16/Nov/2020 11:45:32 [NetworkInterface.cpp:2596] Started flow user script hooks loop on interface /dev/fd/63 [id: 0]...
16/Nov/2020 11:45:32 [main.cpp:382] Working directory: /var/lib/ntopng
16/Nov/2020 11:45:32 [main.cpp:384] Scripts/HTML pages directory: /usr/share/ntopng
16/Nov/2020 11:45:32 [Ntop.cpp:440] Welcome to ntopng x86_64 v.4.2.201112 - (C) 1998-20 ntop.org
16/Nov/2020 11:45:32 [NtopPro.cpp:701] [LICENSE] System Id:	BE150604920BAB21
16/Nov/2020 11:45:32 [NtopPro.cpp:702] [LICENSE] Edition:	Enterprise L
16/Nov/2020 11:45:32 [NtopPro.cpp:703] [LICENSE] License Type:	Time-Limited [Empty license file] License
16/Nov/2020 11:45:32 [NtopPro.cpp:723] [LICENSE] Validity:	Until Mon Nov 16 11:55:32 2020
16/Nov/2020 11:45:32 [PeriodicActivities.cpp:109] Started periodic activities loop...
16/Nov/2020 11:45:33 [startup.lua:21] [user_scripts.lua:772] Script 'attack_mitigation_snmp' configuration is missing from subdir 'host'. New user script?
16/Nov/2020 11:45:33 [startup.lua:21] [user_scripts.lua:772] Script 'device_connection_disconnection' configuration is missing from subdir 'interface'. New user script?
16/Nov/2020 11:45:33 [startup.lua:21] [user_scripts.lua:772] Script 'lateral_movement' configuration is missing from subdir 'interface'. New user script?
16/Nov/2020 11:45:33 [startup.lua:21] [user_scripts.lua:772] Script 'pool_connection_disconnection' configuration is missing from subdir 'interface'. New user script?
16/Nov/2020 11:45:33 [startup.lua:21] [user_scripts.lua:772] Script 'pool_quota_exceeded' configuration is missing from subdir 'interface'. New user script?
16/Nov/2020 11:45:33 [startup.lua:21] [user_scripts.lua:772] Script 'lldp_topology_changed' configuration is missing from subdir 'snmp_device'. New user script?
16/Nov/2020 11:45:33 [startup.lua:21] [user_scripts.lua:772] Script 'flow_risks' configuration is missing from subdir 'flow'. New user script?
16/Nov/2020 11:45:33 [startup.lua:21] [user_scripts.lua:772] Script 'iec60870_5_104' configuration is missing from subdir 'flow'. New user script?
16/Nov/2020 11:45:33 [startup.lua:21] [user_scripts.lua:772] Script 'tcp_issues_generic' configuration is missing from subdir 'flow'. New user script?
16/Nov/2020 11:45:33 [startup.lua:21] [user_scripts.lua:772] Script 'tcp_issues_packets' configuration is missing from subdir 'flow'. New user script?
16/Nov/2020 11:45:33 [startup.lua:21] [user_scripts.lua:772] Script 'tls_certificate_selfsigned' configuration is missing from subdir 'flow'. New user script?
16/Nov/2020 11:45:33 [startup.lua:21] [user_scripts.lua:772] Script 'unexpected_dhcp' configuration is missing from subdir 'flow'. New user script?
16/Nov/2020 11:45:33 [startup.lua:21] [user_scripts.lua:772] Script 'unexpected_dns' configuration is missing from subdir 'flow'. New user script?
16/Nov/2020 11:45:33 [startup.lua:21] [user_scripts.lua:772] Script 'unexpected_ntp' configuration is missing from subdir 'flow'. New user script?
16/Nov/2020 11:45:33 [startup.lua:21] [user_scripts.lua:772] Script 'unexpected_smtp' configuration is missing from subdir 'flow'. New user script?
16/Nov/2020 11:45:33 [startup.lua:21] [user_scripts.lua:772] Script 'active_monitoring' configuration is missing from subdir 'system'. New user script?
16/Nov/2020 11:45:33 [startup.lua:21] [user_scripts.lua:772] Script 'ip_reassignment' configuration is missing from subdir 'system'. New user script?
16/Nov/2020 11:45:33 [startup.lua:21] [user_scripts.lua:772] Script 'remote_to_remote' configuration is missing from subdir 'system'. New user script?
16/Nov/2020 11:45:33 [startup.lua:21] [user_scripts.lua:772] Script 'fortinet' configuration is missing from subdir 'syslog'. New user script?
16/Nov/2020 11:45:33 [startup.lua:21] [user_scripts.lua:772] Script 'host_log' configuration is missing from subdir 'syslog'. New user script?
16/Nov/2020 11:45:33 [startup.lua:21] [user_scripts.lua:772] Script 'openvpn' configuration is missing from subdir 'syslog'. New user script?
16/Nov/2020 11:45:33 [startup.lua:21] [user_scripts.lua:772] Script 'opnsense' configuration is missing from subdir 'syslog'. New user script?
16/Nov/2020 11:45:33 [startup.lua:21] [user_scripts.lua:772] Script 'sonicwall' configuration is missing from subdir 'syslog'. New user script?
16/Nov/2020 11:45:33 [startup.lua:21] [user_scripts.lua:772] Script 'sophos' configuration is missing from subdir 'syslog'. New user script?
16/Nov/2020 11:45:33 [startup.lua:50] Processing startup.lua: please hold on...
16/Nov/2020 11:45:33 [startup.lua:144] [lists_utils.lua:758] Refreshing category lists...
16/Nov/2020 11:45:35 [startup.lua:144] [lists_utils.lua:411] Updating list 'Snort IP Blacklist' [https://snort.org/documents/ip-blacklist]... OK
16/Nov/2020 11:45:35 [startup.lua:144] [lists_utils.lua:696] Category Lists (694 hosts, 4195 IPs, 98 JA3) loaded in 0 sec
16/Nov/2020 11:45:35 [startup.lua:203] Loading category lists...
16/Nov/2020 11:45:35 [startup.lua:205] Loading category lists done
16/Nov/2020 11:45:35 [startup.lua:214] Running /usr/share/ntopng/local/scripts/callbacks/system/startup.lua
16/Nov/2020 11:45:35 [startup.lua:215] [startup.lua:40] API user already existing...
16/Nov/2020 11:45:35 [startup.lua:218] Startup completed: ntopng is now operational
16/Nov/2020 11:45:35 [PeriodicActivities.cpp:172] Each periodic activity script will use 2 threads
16/Nov/2020 11:45:35 [NetworkInterface.cpp:2738] Started packet polling on interface /dev/fd/63 [id: 0]...
16/Nov/2020 11:45:36 [PcapInterface.cpp:226] No more pcap files to read
16/Nov/2020 11:45:36 [PcapInterface.cpp:336] Terminated packet polling for /dev/fd/63
16/Nov/2020 11:45:36 [NetworkInterface.cpp:2624] Flow dump thread completed for /dev/fd/63
16/Nov/2020 11:45:39 [Ntop.cpp:2540] Terminating periodic activities
```

**Comment by ValeDaRold** [comment]
Hi @cardigliano and @lucaderi, 
Thanks for the quick reply and the fix, we tried it out the nightly build and it is solving our problem. Is it possible to have the same fix also in the stable 4.2 release?
Thanks in advance.

**Comment by cardigliano** [comment]
@ValeDaRold I backported it, it will be available in the packages later today
